src/main/resources/include/hail/Region.h

```diff
@@ -0,0 +1,176 @@
+#ifndef HAIL_REGION_H
+#define HAIL_REGION_H 1
+
+#include "hail/NativeObj.h"
+#include "hail/NativePtr.h"
+#include <cstdlib>
+#include <cstring>
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+namespace hail {
+
+class Region;
+using RegionPtr = std::shared_ptr<Region>;
+
+class Region : public NativeObj {
+private:
+  static constexpr long kChunkCap = 64*1024;
+  static constexpr long kMaxSmall = 1024;
+  static constexpr size_t kNumBigToKeep = 4;
+  
+  struct BigAlloc {
+    char* buf_;
+    long  size_;
+    
+    inline BigAlloc() : buf_(nullptr), size_(0) { }
+
+    inline BigAlloc(char* buf, long size) : buf_(buf), size_(size) { }
+    
+    inline BigAlloc(const BigAlloc& b) : buf_(b.buf_), size_(b.size_) { }
+    
+    inline BigAlloc& operator=(const BigAlloc& b) {
+      buf_ = b.buf_;
+      size_ = b.size_;
+      return(*this);
+    }
+  };
+
+public:
+  char* chunk_;
+  long  pos_;
+  std::vector<char*> free_chunks_;
+  std::vector<char*> full_chunks_;
+  std::vector<BigAlloc> big_free_;
+  std::vector<BigAlloc> big_used_;
+  std::vector<RegionPtr> required_regions_;
+
+public:  
+  Region() :
+    chunk_(nullptr),
+    pos_(kChunkCap) {
+  }
+  
+  virtual ~Region() {
+    if (chunk_) free(chunk_);
+    for (auto p : free_chunks_) free(p);
+    for (auto p : full_chunks_) free(p);
+    for (auto& b : big_free_) free(b.buf_);
+    for (auto& b : big_used_) free(b.buf_);
+  }
+  
+  // clear_but_keep_mem() will make the Region empty without free'ing chunks
+  void clear_but_keep_mem() {
+    pos_ = (chunk_ ? 0 : kChunkCap);
+    for (auto p : full_chunks_) free_chunks_.push_back(p);
+    full_chunks_.clear();
+    // Move big_used_ to big_free_
+    for (auto& used : big_used_) {
+      big_free_.emplace_back(used.buf_, used.size_);
+      used.buf_ = nullptr;
+    }
+    big_used_.clear();
+    // Sort in descending order of size
+    std::sort(
```

Comment:
My reasoning is that a very large part of the Hail data is going to be the array of entries,
which will typically be 10K to 500K structs.  If you aren't going to optimize the re-use of those
(few, often only one) large allocations, then you aren't really re-using the memory allocation.

Reply:
> then you aren't really re-using the memory allocation

Why is this desirable?  See my other comment, Region was intended to optimize fast path allocation and make free very cheap.  Calling malloc/free ~once/row doesn't appear to violate either goal.  In fact, it was your malloc benchmark was what convinced me of this!  Also, I'm not convinced malloc won't have roughly similar behavior: if you're freeing and allocating similar sized objects, malloc will return one (or a few) of the same memory blocks each time (depending of course on what other intermediate allocations are going on).

I'm trying out the idea that moving these big allocations using std::move is another hint that we don't even want to be doing this in the first place.

If you still think this will make a significant difference, I defer to you and let's leave it in.